skrub/_table_vectorizer.py

```diff
@@ -495,147 +435,101 @@ def _clone_transformers(self):
             self.n_jobs,
         )
 
-    def _auto_cast(self, X: pd.DataFrame) -> pd.DataFrame:
-        """Takes a dataframe and tries to convert its columns to their best possible
-        data type.
+    def auto_cast(self, X, reset=True):
+        """Convert each column of a dataframe to the "optimal" dtype \
+            for scikit-learn estimators.
+
+        The main heuristics are the following:
+        - pandas extension dtypes conversion to numpy dtype
+        - datetime conversion using ``skrub.to_datetime``
+        - numeric conversion using ``pandas.to_numeric``
+        - numeric columns with missing values are converted to float to input np.nan
+        - categorical columns dtypes are updated with the new entries (if any)
+          during transform.
 
         Parameters
         ----------
         X : :obj:`~pandas.DataFrame` of shape (n_samples, n_features)
             The data to be transformed.
 
+        reset : bool, default=True
+            If set to ``True`` (during fit), creates ``type_per_column_``,
+            the mapping between columns of the training dataframe and their types.
+            If set to ``False`` (during transform), updates ``type_per_column_``
+            for the categorical columns with the new categories seen during transform.
+
         Returns
         -------
-        :obj:`~pandas.DataFrame`
-            The same :obj:`~pandas.DataFrame`, with its columns cast to their
-            best possible data type.
+        X : :obj:`~pandas.DataFrame`
+            The same :obj:`~pandas.DataFrame`, with its columns cast.
         """
-        # Handle missing values
         for col in X.columns:
-            # Convert pandas' NaN value (pd.NA) to numpy NaN value (np.nan)
-            # because the former tends to raise all kind of issues when dealing
-            # with scikit-learn (as of version 0.24).
-            if _has_missing_values(X[col]):
+            # Convert pd.NA to np.nan because the former tends to raise all
```

Comment:
The position of this comment makes it seem like `_replace_false_missing` makes the pd.NA -> np.nan conversion, but that actually happens later. could we move it or reward it ?